nomad/fsm.go

```diff
@@ -578,34 +588,51 @@ func (n *nomadFSM) Restore(old io.ReadCloser) error {
 		}
 	}
 
+	restore.Commit()
+
 	// Create Job Summaries
-	// The entire snapshot has to be restored first before we create the missing
-	// job summaries so that the indexes are updated and we know the highest
-	// index
 	// COMPAT 0.4 -> 0.4.1
-	jobs, err := restore.JobsWithoutSummary()
+	index, err := n.state.Index("job_summary")
 	if err != nil {
-		fmt.Errorf("error retreiving jobs during restore: %v", err)
-	}
-	if err := restore.CreateJobSummaries(jobs); err != nil {
-		return fmt.Errorf("error creating job summaries: %v", err)
+		return fmt.Errorf("couldn't fetch index of job summary table: %v", err)
 	}
 
-	restore.Commit()
+	// If the index is 0 that means there is no job summary in the snapshot so
+	// we will have to create them
+	if index == 0 {
+		// query the latest index
+		latestIndex, err := n.state.LatestIndex()
+		if err != nil {
+			return fmt.Errorf("unable to query latest index: %v", index)
+		}
+		if err := n.state.ReconcileJobSummaries(latestIndex); err != nil {
+			return fmt.Errorf("error reconciling summaries: %v", err)
+		}
+		if err := n.reconcileQueuedAllocations(latestIndex); err != nil {
+			return fmt.Errorf("error re-computing the number of queued allocations:; %v", err)
+		}
+	}
 
-	// Reconciling the queued allocations
-	return n.reconcileSummaries(jobs)
+	return nil
 }
 
 // reconcileSummaries re-calculates the queued allocations for every job that we
 // created a Job Summary during the snap shot restore
-func (n *nomadFSM) reconcileSummaries(jobs []*structs.Job) error {
-	// Start the state restore
-	restore, err := n.state.Restore()
+func (n *nomadFSM) reconcileQueuedAllocations(index uint64) error {
+	// Get all the jobs
+	iter, err := n.state.Jobs()
 	if err != nil {
 		return err
 	}
-	defer restore.Abort()
+	var jobs []*structs.Job
+	for {
+		rawJob := iter.Next()
+		if rawJob == nil {
+			break
+		}
+		jobs = append(jobs, rawJob.(*structs.Job))
+	}
+
 	snap, err := n.state.Snapshot()
 	if err != nil {
 		return fmt.Errorf("unable to create snapshot: %v", err)
```

Comment:
Put a comment on top of the for loop that for every job we dry run the scheduler such that we can repopulate queued